File: platforms/ios/Sources/WhiskerRuntime/module/PendingModuleEvents.swift
import Foundation
import WhiskerModule

struct PendingModuleEvent: Equatable {
    let epoch: UInt64
    let module: String
    let event: String
    let payload: WhiskerValue
}

/// Any-queue ingress buffer drained once per main-loop turn.
final class PendingModuleEvents {
    private let lock = NSLock()
    private var queue: [PendingModuleEvent] = []
    private var flushScheduled = false

    /// Returns true only when the caller must schedule the shared flush.
    func offer(_ event: PendingModuleEvent) -> Bool {
        lock.lock()
        defer { lock.unlock() }
        queue.append(event)
        guard !flushScheduled else { return false }
        flushScheduled = true
        return true
    }

    func drain(epoch: UInt64) -> [PendingModuleEvent] {
        lock.lock()
        defer { lock.unlock() }
        flushScheduled = false
        let current = queue.filter { $0.epoch == epoch }
        queue.removeAll(keepingCapacity: true)
        return current
    }

    func clear() {
        lock.lock()
        queue.removeAll(keepingCapacity: false)
        flushScheduled = false
        lock.unlock()
    }
}

File: platforms/android/runtime/src/main/kotlin/rs/whisker/runtime/module/PendingModuleEvents.kt
package rs.whisker.runtime.module

import rs.whisker.runtime.WhiskerValue

internal data class PendingModuleEvent(
    val epoch: Long,
    val module: String,
    val event: String,
    val payload: WhiskerValue,
)

/** Any-thread ingress queue drained once per main-loop turn. */
internal class PendingModuleEvents {
    private val lock = Any()
    private val queue = ArrayDeque<PendingModuleEvent>()
    private var flushScheduled = false

    /** Returns true only when the caller must schedule the shared flush. */
    fun offer(event: PendingModuleEvent): Boolean = synchronized(lock) {
        queue.addLast(event)
        if (flushScheduled) {
            false
        } else {
            flushScheduled = true
            true
        }
    }

    fun drain(epoch: Long): List<PendingModuleEvent> = synchronized(lock) {
        flushScheduled = false
        buildList {
            while (queue.isNotEmpty()) {
                queue.removeFirst().takeIf { it.epoch == epoch }?.let(::add)
            }
        }
    }

    fun clear() = synchronized(lock) {
        queue.clear()
        flushScheduled = false
    }
}
